how-to/workspace-platform-starter/client/src/framework/workspace/notification-client.ts

```diff
@@ -0,0 +1,211 @@
+	 * @param options the notification to update.
+	 * @returns the updated notification.
+	 */
+	public async update<T extends UpdatableNotificationOptions>(options: T): Promise<Notification> {
+		if (!this.hasId(options.id)) {
+			options.id = this.createId(options.id);
+		}
+
+		return notificationUpdate(options);
+	}
+
+	/**
+	 * Get all notifications.
+	 * @returns all the notifications.
+	 */
+	public async getAll(): Promise<Notification[]> {
+		const notifications = await notificationGetAll();
+		return notifications.filter((entry) => entry.id.startsWith(this._idPrefix));
+	}
+
+	/**
+	 * Clears a specific notification.
+	 * @param id the id of the notification to clear.
+	 * @returns whether the notification was cleared.
+	 */
+	public async clear(id: string): Promise<boolean> {
+		const notificationToClear = this.createId(id);
+		return notificationClear(notificationToClear);
+	}
+
+	/**
+	 * Clear all notifications.
+	 * @returns the number of notifications that was cleared.
+	 */
+	public async clearAll(): Promise<number> {
+		const notifications = await this.getAll();
+		let count = 0;
+		for (const notification of notifications) {
+			const removed = await this.clear(notification.id);
+			if (removed) {
+				count++;
+			}
+		}
+		return count;
+	}
+
+	/**
+	 * Shows the notification center.
+	 * @param options the options to show.
+	 */
+	public async show(options?: ShowOptions | undefined): Promise<void> {
+		await notificationShow(options);
+	}
+
+	/**
+	 * The ability to hide the notification center.
+	 */
+	public async hide(): Promise<void> {
+		await notificationHide();
+	}
+
+	// public async addEventListener(eventType: "notification-form-submitted", listener: (event: NotificationFormSubmittedEvent) => void): Promise<void>;
+	// public async addEventListener(eventType: "notification-action", listener: (event: NotificationActionEvent) => void): Promise<void>;
+	// public async addEventListener(eventType: "notification-created", listener: (event: NotificationCreatedEvent) => void): Promise<void>;
+	// public async addEventListener(eventType: "notification-toast-dismissed", listener: (event: NotificationToastDismissedEvent) => void): Promise<void>;
+	// public async addEventListener(eventType: "notification-closed", listener: (event: NotificationClosedEvent) => void): Promise<void>;
+	// public async addEventListener(eventType: "notification-reminder-created", listener: (event: NotificationReminderCreatedEvent) => void): Promise<void>;
+	// public async addEventListener(eventType: "notification-reminder-removed", listener: (event: NotificationReminderRemovedEvent) => void): Promise<void>;
+	// public async addEventListener(eventType: "notification-count-changed", listener: (event: NotificationsCountChanged) => void): Promise<void>;
+	/**
+	 * Tells you when a notification was created.
+	 * @param eventType the type of event to listen to
+	 * @param listener the listener.
+	 */
+	public async addEventListener(
```

Comment:
To better handle the multiple event types add
```
/**
 * Mapping of all notification events.
 */
export interface NotificationsEventMap {
	"notification-created": NotificationCreatedEvent;
	"notification-action": NotificationActionEvent;
	"notification-form-submitted": NotificationFormSubmittedEvent;
	"notification-toast-dismissed": NotificationToastDismissedEvent;
	"notification-closed": NotificationClosedEvent;
	"notification-reminder-created": NotificationReminderCreatedEvent;
	"notification-reminder-removed": NotificationReminderRemovedEvent;
	"notification-count-changed": NotificationsCountChanged;
}
```
Update sigs to
```
addEventListener<K extends keyof NotificationsEventMap>(
		eventType: K,
		listener: (event: NotificationsEventMap[K]) => void
	): Promise<void>;

	/**
	 *
	 */
	removeEventListener<K extends keyof NotificationsEventMap>(
		eventType: K,
		listener: (event: NotificationsEventMap[K]) => void
	): Promise<void>;
```

Update listeners to
```
private _listenerRegister: {
		id: string;
		listener: (event: NotificationsEventMap[keyof NotificationsEventMap]) => void;
		wrappedListener: (event: NotificationsEventMap[keyof NotificationsEventMap]) => void;
	}[] = [];
```

Update implementations to

```
public async addEventListener<K extends keyof NotificationsEventMap>(
		eventType: K,
		listener: (event: NotificationsEventMap[K]) => void
	): Promise<void> {
		const mappedListener: {
			id: string;
			listener: (event: NotificationsEventMap[K]) => void;
			wrappedListener: (event: NotificationsEventMap[K]) => void;
		} = {
			id: randomUUID(),
			listener,
			wrappedListener: (event: NotificationsEventMap[K]) => {
				if ("notification" in event && this.hasId(event.notification.id)) {
					listener(event);
				} else {
					// What do we do with events without a notification like count-changed ?
				}
			}
		};
		this._listenerRegister.push(
			mappedListener as {
				id: string;
				listener: (event: NotificationsEventMap[keyof NotificationsEventMap]) => void;
				wrappedListener: (event: NotificationsEventMap[keyof NotificationsEventMap]) => void;
			}
		);
		await notificationAddEventListener(
			eventType as unknown as any,
			mappedListener.wrappedListener as unknown as any
		);
	}

	/**
	 * Removes an event listener.
	 * @param eventType the type of event to listen to
	 * @param listener the listener.
	 */
	public async removeEventListener<K extends keyof NotificationsEventMap>(
		eventType: K,
		listener: (event: NotificationsEventMap[K]) => void
	): Promise<void> {
		// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
		const mappedListener = this._listenerRegister.find((entry) => entry.listener === listener);
		if (!isEmpty(mappedListener)) {
			await notificationRemoveEventListener(eventType as unknown as any, mappedListener.wrappedListener);
			this._listenerRegister = this._listenerRegister.filter((entry) => entry.id !== mappedListener.id);
		}
	}
```

Reply:
Really helpful!

Used as a base for the updates.